enhancements/two-nodes-openshift/2no.md

```diff
@@ -0,0 +1,516 @@
+The primary exception is (re)booting one of the peers while the other is offline, and expected to remain so.
+
+As in a 3-node control-plane cluster, starting only one node is not expected to result in a functioning cluster.
+Should the admin wish for the control-plane to start, the admin will need to execute a supplied confirmation command on the active cluster node. 
+This command will grant quorum to the RHEL-HA components, authorizing it to fence it's peer and start etcd in as a cluster-of-one read/write mode.
+Confirmation can be given at any point and optionally make use of SSH to facilitate initiation by an external script.
+
+### API Extensions
+
+There are two related but ultimately orthogonal capabilities that may require API extensions.
+
+1. Identify the cluster as having a unique topology
+2. Tell CEO when it is safe for it to disable certain membership related functionalities
+
+#### Unique Topology
+
+A mechanism is needed for the installer and other components to understand that this is a 2 node control-plane topology which may require different handling.
+
+TODO: pros and cons of creating a new PlatformType, vs. feature gate, vs adding a new field to `PlatformSpec` or `BareMetalPlatformSpec`
+
+#### CEO Trigger
+
+Initially the creation of an etcd cluster will be driven in the same way as other platforms.
+Once the cluster has two members, the etcd daemon will be removed from the static pod definition and recreated as a resource controlled by RHEL-HA.
+At this point, the Cluster Etcd Operator (CEO) will be made aware of this change so that some membership management functionality that is now handled by RHEL-HA can be disabled.
+This will be achieved by having the same entity that drives the configuration of RHEL-HA use the OpenShift API to update a field in the `BareMetalPlatformSpec` portion of the `Infrastructure` CR - which can only succeed if the control-plane is healthy.
+
+To enable this flow, we propose the addition of a `externallyManagedEtcd` field to the `BareMetalPlatformSpec` which defaults to False.
+This will limit the scope of CEO behavioural changes to that specific platform, and well as allow the use of a tightly scoped credential to make the change.
+An alternative being to grant write access to all `ConfigMaps` in the `openshift-config` namespace.
+
+### Topology Considerations
+
+2NO represents a new topology, and is not appropriate for use with HyperShift, SNO, or MicroShift
+
+#### Standalone Clusters
+
+Is the change relevant for standalone clusters?
+TODO: Exactly what is the definition of a standalone cluster?  Disconnected?  Physical hardware?
+
+### Implementation Details/Notes/Constraints
+
+While the target installation requires exactly 2 nodes, this will be achieved by building support in the core installer for a "bootstrap plus 2 nodes" flow, and then using Assisted Installer's ability to bootstrap-in-place to remove the requirement for a bootstrap node.
```

Comment:
I don't think we have clarity on this point yet.

'bootstrap-in-place' in the installer refers to some stuff that is used exclusively for SNO (by assisted).
When installing non-SNO clusters, assisted does the bootstrapping in-place on one of the control plane nodes, but not using the 'bootstrap-in-place' parts of the installer.

Anyway, I don't see any point in implementing this twice.

Reply:
I don't believe there is any plan to implement this again, but to reuse Assisted's existing functionality.
Are you suggesting this is not possible?

Reply:
Sorry, I was unclear.

I don't think we should implement bootstrapping of 2-node twice, once for separate bootstrap node and once for in-place bootstrap. Just do the second and don't worry about IPI support.

Reply:
@beekhof sorry if my input in the meetings was confusing. Whatever way the team wants to implemented 2-node install is fine as long as it works using the openshift-installer in isolation.

I took to referring to "using a bootstrap node" as a synonym for "not assisted installer", but that's not entirely true.